tests/compile-test.rs

```diff
@@ -46,9 +46,33 @@ fn config(mode: &str, dir: PathBuf) -> compiletest::Config {
         config.run_lib_path = rustc_lib_path();
         config.compile_lib_path = rustc_lib_path();
     }
+
+    // When we'll want to use `extern crate ..` for a dependency that is used
+    // both by the crate and the compiler itself, we can't simply pass -L flags
+    // as we'll get a duplicate matching versions. Instead, disambiguate with
+    // `--extern dep=path`.
+    // See https://github.com/rust-lang/rust-clippy/issues/4015.
+    let needs_disambiguation = ["serde"];
+    // This assumes that deps are compiled (they are for Cargo integration tests).
+    let deps = std::fs::read_dir(host_libs().join("deps")).unwrap();
+    let disambiguated = deps
+        .filter_map(|dep| dep.ok().map(|x| x.path()))
+        .filter(|path| path.file_name().is_some())
+        .filter_map(|path| {
+            let name = path.file_name().unwrap().to_string_lossy();
```

Comment:
Any idea how to avoid the `is_some()`/`unwrap()` dance?
I tried naively to do
```rust
filter_map(|path| path.file_name().map(move |name| (path, name)|)
```
before but then I get lifetime errors because I'm returning `(PathBuf, &OsStr)` which borrows from it; ideally I'd just borrow everything for the same lifetime and be done with it

Reply:
The problem is the fact that you are trying to make the references live across multiple `filter_*` calls. If you use a single call, it works out:

```rust
    let disambiguated = deps
        .filter_map(|dep| {
            let path = dep.ok()?.path();
            let name = path.file_name()?.to_string_lossy();
            // NOTE: This only handles a single dep
            // https://github.com/laumann/compiletest-rs/issues/101
            needs_disambiguation
                .iter()
                .find(|dep| name.starts_with(&format!("lib{}-", dep)))
                .map(|dep| format!("--extern {}={}", dep, path.display()))
        })
        .collect::<Vec<_>>();
```

Reply:
Ah, I forgot that question mark works at the closure boundaries! Nice, let me update this real quick